tex/testing.tex

```diff
@@ -0,0 +1,33 @@
+\chapter{Testing}
+Due to Extend being a large undertaking, we took steps to ensure that all features were working as the design of the language intended.
+
+\medskip \noindent
+This was done through implementing test cases that isolated specific aspects of the Extend language to ensure that each feature worked correctly. For basic components, we wrote a plethora of tests to illustrate functionality. For undertakings that required more debate on the design of the language, other tests were created and modified throughout development.
+
+\section{Feature Integration \& Testing}
+Development of new features naturally means that they must be deemed legal by the scanner, parser, semantic analyzer, and code generator. As we developed new features, the process was roughly as follows:
+  \begin{enumerate}
+    \item Write a simple test that illustrated the feature to test.
+    \item Write the expected output of the aforementioned test to a text file.
+    \item Confirm that the scanner consumes the tokens related to the feature.
+    \item Confirm that the parser grammar has been adjusted to accomodate the new feature.
+    \item Confirm that the semantic analyzer and transformer can properly identify and check the new feature code.
+    \item Confirm that code generation generates the appropriate LLVM IR for the new features - such as allocating memory, building calls, and more.
+    \item Ensure that the test written can write its output to stdout, to be compared with expected output.
+    \item Compile and test the code to ensure that the code has worked to the team's expectations.
+  \end{enumerate}
+
+  \medskip \noindent
+  Earlier in the development process, we tested the front end of our compiler by JSON-ifying the abstract syntax tree and examinining it. As we settled into full-fledged development, we would test with a full-feature regression test suite.
+
+\section{Regression Test Suite}
+Extend's regression test suite is executable through the \texttt{testscript.sh} script at the top level of the project. There are over 100 integration test files for various features of the Extend language, and a corresponding file with their expected output to \texttt{stdout}. This is to ensure that the successful implementation of one feature does not impact that of others.
+
+\medskip \noindent
+The test script iterates through every test file in the \texttt{inputs\_regression} directory. It compiles and executes the test, and compares it with the corresponding expected output file living in the \texttt{expected} directory. Some of these tests were "passing" tests, meant to confirm successful output, and some were "failing" tests, identifying the ability of the language to detect illegal actions and exit gracefully.
```

Comment:
Not correct. Regression tests were placed in inputs_regression, not passing tests were placed in inputs. That way we could see what was still left to do. Whenever a new test passed it automatically moved over.